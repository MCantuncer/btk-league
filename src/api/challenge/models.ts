import { Field, InputType } from 'type-graphql';
import { BaseInputWithDate } from '../../models/common';

@InputType()
export class ChallengeResponseInput extends BaseInputWithDate {
  @Field(() => Boolean)
  accepted: boolean;
}
