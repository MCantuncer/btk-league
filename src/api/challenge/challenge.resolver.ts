// import {Challenge, ChallengeModel} from "../../models/challenge/entity";
// import {Ctx, Query, Resolver} from "type-graphql";
//
// @Resolver(() => Challenge)
// export default class {
//   @Query(() => Challenge[])
//   async challengeList(@Ctx() ctx): Promise<Challenge[] | null> {
//     const challengeList = ChallengeModel.find({ challenged: ctx.user})
//   }
// }
//
