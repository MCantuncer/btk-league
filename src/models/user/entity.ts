import { Field, Int, ObjectType } from 'type-graphql';
import { getModelForClass, prop } from '@typegoose/typegoose';
import { UserTypeEnum } from './enums';
import { EntityWithIdAndDates } from '../base';

@ObjectType()
export class User extends EntityWithIdAndDates {
  @prop()
  @Field(() => UserTypeEnum)
  userType: UserTypeEnum;

  @prop()
  @Field(() => String)
  firstName: string;

  @prop()
  @Field(() => String)
  lastName: string;

  @prop()
  @Field(() => String)
  email: string;

  @prop()
  @Field(() => String)
  password: string;

  @prop()
  @Field(() => Int, { nullable: true })
  standing?: number | null;
}

export const UserModel = getModelForClass(User, {
  schemaOptions: { collection: 'users' },
});
