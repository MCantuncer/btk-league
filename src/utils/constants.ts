import { CookieOptions } from 'express';

export const SESSION_COOKIE_NAME = '_session_identity';
export const COOKIE_OPTIONS: CookieOptions = { httpOnly: true, sameSite: true };

export interface ITokenData {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  iat?: number;
  exp?: number;
  sub?: string;
  aud?: string;
  iss?: string;
}

export interface IPasswordResetTokenData {
  email: string;
  iat?: number;
  exp?: number;
  sub?: string;
  aud?: string;
  iss?: string;
}
