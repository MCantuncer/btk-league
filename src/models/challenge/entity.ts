import { Field, ObjectType } from 'type-graphql';
import { EntityWithIdAndDates } from '../base';
import { getModelForClass, prop, Ref } from '@typegoose/typegoose';
import { User } from '../user/entity';

@ObjectType()
export class Challenge extends EntityWithIdAndDates {
  @prop({ ref: User, type: String })
  @Field(() => String)
  challenger: Ref<User>;

  @prop({ ref: User, type: String })
  @Field(() => String)
  challenged: Ref<User>;

  @prop()
  @Field(() => Boolean, { nullable: true })
  accepted?: boolean | null;
}

export const ChallengeModel = getModelForClass(Challenge, {
  schemaOptions: { collection: 'challenges' },
});
