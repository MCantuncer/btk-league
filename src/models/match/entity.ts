import { Field, GraphQLTimestamp, Int, ObjectType } from 'type-graphql';
import { EntityWithIdAndDates } from '../base';
import { getModelForClass, modelOptions, prop, Ref } from '@typegoose/typegoose';
import { User } from '../user/entity';
import { WinnerEnum } from './enums';

@ObjectType()
@modelOptions({ schemaOptions: { _id: false } })
export class SetResult {
  @prop()
  @Field(() => Int)
  homeGame: number;

  @prop()
  @Field(() => Int)
  visitorGame: number;

  @prop()
  @Field(() => WinnerEnum, { nullable: true })
  winner?: WinnerEnum | null;
}

@ObjectType()
export class Match extends EntityWithIdAndDates {
  @prop({ ref: User, type: String })
  @Field(() => String)
  home: Ref<User>;

  @prop({ ref: User, type: String })
  @Field(() => String)
  visitor: Ref<User>;

  @prop()
  @Field(() => [SetResult], { nullable: true })
  result?: SetResult[] | null;

  @prop()
  @Field(() => WinnerEnum, { nullable: true })
  matchWinner?: WinnerEnum | null;

  @prop()
  @Field(() => GraphQLTimestamp, { nullable: true })
  matchDate?: Date | null;
}

export const MatchModel = getModelForClass(Match, {
  schemaOptions: { collection: 'matches' },
});
