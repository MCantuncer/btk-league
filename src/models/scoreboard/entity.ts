import { Field, Int, ObjectType } from 'type-graphql';
import { prop, Ref } from '@typegoose/typegoose';
import { User } from '../user/entity';

@ObjectType()
export class ScoreBoard {
  @prop({ ref: User, type: String })
  @Field(() => String)
  user: Ref<User>;

  @prop()
  @Field(() => Int)
  points: number;
}
