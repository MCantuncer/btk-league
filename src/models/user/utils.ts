import { User } from './entity';
import { hash } from 'argon2';

export class UserUtils {
  user: User;
  constructor(user: User) {
    this.user = user;
  }

  setPassword = async (password: string) => (this.user.password = await hash(password));
}
