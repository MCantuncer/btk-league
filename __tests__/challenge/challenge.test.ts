import { testMutate, testQuery } from '../../jest.setup';
import { UserInputs, UserMutations } from '../user/globals';
import { ChallengeInputs, ChallengeMutations, ChallengeQueries } from './globals';
import { Challenge, ChallengeModel } from '../../src/models/challenge/entity';
import { TEST_USER_ID } from '../../src/utils/constants';

describe('Challenge Tests', () => {
  it('Make a Challenge', async () => {
    let graphqlResponse = await testMutate(UserMutations.Instance.REGISTER, {
      input: { ...UserInputs.Instance().REGISTER_INPUT_1, email: 'tuncer@btk.com' },
    });

    expect(graphqlResponse.errors).toBeUndefined();

    graphqlResponse = await testMutate(ChallengeMutations.Instance.MAKE_CHALLENGE, {
      input: ChallengeInputs.Instance().MAKE_CHALLENGE_INPUT_1,
    });

    expect(graphqlResponse.errors).toBeUndefined();

    const response = (graphqlResponse.data || {}) as { makeChallenge: Challenge };

    expect(response.makeChallenge.challenger).toEqual(TEST_USER_ID);
    expect(response.makeChallenge.challenged).toEqual(UserInputs.CREATE_ID_1);
    expect(response.makeChallenge.accepted).toBeNull();

    expect(response).toMatchSnapshot();
  });

  it('Make another Challenge', async () => {
    let graphqlResponse = await testMutate(UserMutations.Instance.REGISTER, {
      input: { ...UserInputs.Instance().REGISTER_INPUT_1, id: UserInputs.CREATE_ID_2, email: 'can@btk.com' },
    });

    expect(graphqlResponse.errors).toBeUndefined();

    graphqlResponse = await testMutate(ChallengeMutations.Instance.MAKE_CHALLENGE, {
      input: ChallengeInputs.Instance().MAKE_CHALLENGE_INPUT_2,
    });

    expect(graphqlResponse.errors).toBeUndefined();

    const response = (graphqlResponse.data || {}) as { makeChallenge: Challenge };

    expect(response.makeChallenge.challenger).toEqual(TEST_USER_ID);
    expect(response.makeChallenge.challenged).toEqual(UserInputs.CREATE_ID_2);
    expect(response.makeChallenge.accepted).toBeNull();

    expect(response).toMatchSnapshot();
  });

  it('List Challenges that User Challenged', async () => {
    const graphqlResponse = await testQuery(ChallengeQueries.Instance.CHALLENGED_LIST, {});

    expect(graphqlResponse.errors).toBeUndefined();

    const response = (graphqlResponse.data || {}) as { challengedList: Challenge[] };

    expect(response.challengedList.length).toEqual(2);
    expect(response.challengedList[0].challenger).toEqual(TEST_USER_ID);
    expect(response.challengedList[1].challenger).toEqual(TEST_USER_ID);

    expect(response).toMatchSnapshot();
  });

  it('List Challenges that User is Challenged By', async () => {
    let graphqlResponse = await testMutate(UserMutations.Instance.REGISTER, {
      input: { ...UserInputs.Instance().REGISTER_INPUT_1, id: UserInputs.CREATE_ID_3, email: 'mcan@btk.com' },
    });

    expect(graphqlResponse.errors).toBeUndefined();

    await ChallengeModel.create(
      new ChallengeModel({
        _id: ChallengeInputs.CREATE_ID_3,
        challenger: UserInputs.CREATE_ID_3,
        challenged: TEST_USER_ID,
      }),
    );

    graphqlResponse = await testQuery(ChallengeQueries.Instance.CHALLENGED_BY_LIST, {});

    expect(graphqlResponse.errors).toBeUndefined();

    const response = (graphqlResponse.data || {}) as { challengedByList: Challenge[] };

    expect(response.challengedByList).toHaveLength(1);

    expect(response.challengedByList[0].challenger).toEqual(UserInputs.CREATE_ID_3);
    expect(response.challengedByList[0].challenged).toEqual(TEST_USER_ID);

    expect(response).toMatchSnapshot();
  });

  it('Accept a Challenge', async () => {
    const graphqlResponse = await testMutate(ChallengeMutations.Instance.RESPOND_CHALLENGE, {
      input: ChallengeInputs.Instance().RESPOND_CHALLENGE_INPUT_1,
    });

    expect(graphqlResponse.errors).toBeUndefined();

    const response = (graphqlResponse.data || {}) as { respondChallenge: Challenge };

    expect(response.respondChallenge.challenger).toEqual(UserInputs.CREATE_ID_3);
    expect(response.respondChallenge.challenged).toEqual(TEST_USER_ID);
    expect(response.respondChallenge.accepted).toBe(true);

    expect(response).toMatchSnapshot();
  });
});
