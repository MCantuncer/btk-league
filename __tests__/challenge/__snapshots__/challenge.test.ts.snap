// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Challenge Tests Accept a Challenge 1`] = `
Object {
  "respondChallenge": Object {
    "accepted": true,
    "challenged": "ebc44860-e523-45b8-b165-de1b001ae8df",
    "challenger": "dfc9cac2-4240-43d1-9274-1dde9e4a6b18",
    "deleted": false,
    "id": "b5f6b1da-aefd-4c0f-86d9-3551503aea83",
  },
}
`;

exports[`Challenge Tests List Challenges that User Challenged 1`] = `
Object {
  "challengedList": Array [
    Object {
      "accepted": null,
      "challenged": "0bbdcc9f-a146-4d1b-a2bc-65257dadbd48",
      "challenger": "ebc44860-e523-45b8-b165-de1b001ae8df",
      "deleted": false,
      "id": "bfa762a8-e4ad-456f-8de9-2c59c753a3c8",
    },
    Object {
      "accepted": null,
      "challenged": "05837b13-6c10-45f8-9ed5-a88fa6ef9ff2",
      "challenger": "ebc44860-e523-45b8-b165-de1b001ae8df",
      "deleted": false,
      "id": "6592622d-bb40-497a-b92c-452bdaa4320a",
    },
  ],
}
`;

exports[`Challenge Tests List Challenges that User is Challenged By 1`] = `
Object {
  "challengedByList": Array [
    Object {
      "accepted": null,
      "challenged": "ebc44860-e523-45b8-b165-de1b001ae8df",
      "challenger": "dfc9cac2-4240-43d1-9274-1dde9e4a6b18",
      "deleted": false,
      "id": "b5f6b1da-aefd-4c0f-86d9-3551503aea83",
    },
  ],
}
`;

exports[`Challenge Tests Make a Challenge 1`] = `
Object {
  "makeChallenge": Object {
    "accepted": null,
    "challenged": "0bbdcc9f-a146-4d1b-a2bc-65257dadbd48",
    "challenger": "ebc44860-e523-45b8-b165-de1b001ae8df",
    "deleted": false,
    "id": "bfa762a8-e4ad-456f-8de9-2c59c753a3c8",
  },
}
`;

exports[`Challenge Tests Make another Challenge 1`] = `
Object {
  "makeChallenge": Object {
    "accepted": null,
    "challenged": "05837b13-6c10-45f8-9ed5-a88fa6ef9ff2",
    "challenger": "ebc44860-e523-45b8-b165-de1b001ae8df",
    "deleted": false,
    "id": "6592622d-bb40-497a-b92c-452bdaa4320a",
  },
}
`;
